plugins/org.eclipse.emf.ecore/src/org/eclipse/emf/ecore/plugin/EcorePlugin.java

```diff
@@ -1309,4 +1337,361 @@ private static void computeModels(Map<URI, URI> pluginMap, Map<String, URI> nsUR
    * Since 2.14
    */
   public static final String ANNOTATION_VALIDATOR_PPID = "annotation_validator";
+
+  /**
+   * In case we are not running in Eclipse, this BundleActivator will be used.
+   */
+  private static class PlainOSGiBundleActivator implements BundleActivator
+  {
+    private ServiceTracker<EPackage.Registry, Object> osgiEPackageRegistryTracker;
+
+    @Override
+    public void start(BundleContext context) throws Exception
+    {
+      if (defaultRegistryImplementation == null)
+      {
+        OSGiDelegatEPackageRegistry osgiDelegateEPackageRegistry = new OSGiDelegatEPackageRegistry(context);
+        osgiEPackageRegistryTracker = new ServiceTracker<>(context, OSGiDelegatEPackageRegistry.FILTER, osgiDelegateEPackageRegistry);
+        osgiEPackageRegistryTracker.open();
+        defaultRegistryImplementation = osgiDelegateEPackageRegistry;
+      }
+    }
+
+    @Override
+    public void stop(BundleContext context) throws Exception
+    {
+      if (osgiEPackageRegistryTracker != null)
+      {
+        osgiEPackageRegistryTracker.close();
+      }
+    }
+  }
+
+  /**
+   * A {@link org.eclipse.emf.ecore.EPackage.Registry} implementation that looks for an OSGi service that provides a global registry.
+   * Due  to the dynamic nature of OSGi services and the static nature of the default registry
+   * {@link org.eclipse.emf.ecore.EPackage.Registry#INSTANCE} this implementation provides an internal backup registry.
+   * 
+   * The backup is used until the service becomes available.
+   * All write operations will be stored with the backup as well,
+   * so it can restore its role, when the service stops.
+   * When a suitable service becomes available, all {@link EPackage}s will be added to the new service registry.
+   * 
+   * Only services with the property <code>emf.default.epackage.registry=true</code> will be considered.
+   * If multiple such services are available, the one with the highest service rank will be used
+   * or, if no rank is specified, the first one will be used.
+   */
+  private static class OSGiDelegatEPackageRegistry extends AbstractMap<String, Object> implements EPackage.Registry, ServiceTrackerCustomizer<EPackage.Registry, Object>
```

Comment:
Note that I extended AbstractMap so that the equals and hashCode contracts are properly implemented/respected.